jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-prototype.c

```diff
@@ -988,6 +988,86 @@ ecma_builtin_typedarray_prototype_object_to_string (ecma_value_t this_arg) /**<
   return ret_value;
 } /* ecma_builtin_typedarray_prototype_object_to_string */
 
+/**
+ * The %TypedArray%.prototype object's 'subarray' routine.
+ *
+ * See also:
+ *          ES2015, 22.2.3.26
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value.
+ */
+static ecma_value_t
+ecma_builtin_typedarray_prototype_subarray (ecma_value_t this_arg, /**< this argument */
+                                            ecma_value_t begin, /**< begin */
+                                            ecma_value_t end) /**< end */
+{
+  ecma_value_t ret_value = ECMA_VALUE_EMPTY;
+  /* 2.~ 4. */
+  if (!ecma_is_typedarray (this_arg))
+  {
+    return ecma_raise_type_error (ECMA_ERR_MSG ("Argument 'this' is not a TypedArray."));
+  }
+
+  ecma_object_t *src_typedarray_p = ecma_get_object_from_value (this_arg);
+
+  /* 5. buffer */
+  ecma_object_t *src_typedarray_arraybuffer_p = ecma_typedarray_get_arraybuffer (src_typedarray_p);
+
+  /* 6. srcLength */
+  ecma_length_t src_length = ecma_typedarray_get_length (src_typedarray_p);
+
+  /* 9. beginIndex, 12. endIndex */
+  uint32_t begin_index_uint32 = 0, end_index_uint32 = 0;
+
+  /* 7. relativeBegin */
+  ECMA_OP_TO_NUMBER_TRY_CATCH (relative_begin, begin, ret_value);
+  begin_index_uint32 = ecma_builtin_helper_array_index_normalize (relative_begin, src_length);
+  if (ecma_is_value_undefined (end))
+  {
+    end_index_uint32 = (uint32_t) src_length;
+  }
+  else
+  {
+    /* 10. relativeEnd */
+    ECMA_OP_TO_NUMBER_TRY_CATCH (relative_end, end, ret_value);
+
+    end_index_uint32 = ecma_builtin_helper_array_index_normalize (relative_end, src_length);
+
+    ECMA_OP_TO_NUMBER_FINALIZE (relative_end);
+  }
+
+  ECMA_OP_TO_NUMBER_FINALIZE (relative_begin);
+
+  /* 13. newLength */
+  ecma_length_t subarray_length = JERRY_MAX (end_index_uint32 - begin_index_uint32, 0);
+
+  /* 15. elementSize */
+  uint8_t shift = ecma_typedarray_get_element_size_shift (src_typedarray_p);
+  uint8_t element_size = (uint8_t) (1 << shift);
+
+  /* 16. srcByteOffset */
+  ecma_length_t src_byte_offset = ecma_typedarray_get_offset (src_typedarray_p);
+
+  /* 17. beginByteOffset */
+  ecma_length_t begin_byte_offset = src_byte_offset + begin_index_uint32 * element_size;
+
+  uint8_t src_builtin_id = ecma_get_object_builtin_id (src_typedarray_p);
+  ecma_value_t arguments_p[3] =
+  {
+    ecma_make_object_value (src_typedarray_arraybuffer_p),
+    begin_byte_offset,
+    subarray_length
+  };
+
+  if (ecma_is_value_empty (ret_value))
+  {
+    ret_value = ecma_typedarray_helper_dispatch_construct (arguments_p, 3, src_builtin_id);
```

Comment:
On this line I get the following assertion failure:
```
ICE: Assertion 'ecma_typedarray_helper_is_typedarray (builtin_id)' failed at /Users/acalandra/jerryscript/jerry-core/ecma/builtin-objects/typedarray/ecma-builtin-typedarray-helpers.c(ecma_typedarray_helper_dispatch_construct):180.
Error: ERR_FAILED_INTERNAL_ASSERTION
```
Some help solving this issue would be greatly appreciated as I can't seem to figure out why this isn't a typedarray.

Reply:
@AnthonyCalandra The main reason is that `src_builtin_id` is incorrrent and that causes the assertion error.
`uint8_t src_builtin_id = ecma_get_object_builtin_id (src_typedarray_p); ` returns with `ECMA_BUILTIN_ID__COUNT` which means the object is not builtin. 

Currently I cannot find any helper function in the code base to get the `builtin_id` from a `typedarray object`, so my suggestion is to create a function like:

```
static uint8_t
ecma_typedarray_helper_get_builtin_id (ecma_object_t *obj_p) /**< typedarray object **/
{
  #define TYPEDARRAY_ID_CASE(magic_id, builtin_id) \
    case magic_id: \
    { \
      return builtin_id; \
    }

    switch (ecma_object_get_class_name (obj_p))
    {
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_INT8_ARRAY_UL, ECMA_BUILTIN_ID_INT8ARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_UINT8_ARRAY_UL, ECMA_BUILTIN_ID_UINT8ARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_UINT8_CLAMPED_ARRAY_UL, ECMA_BUILTIN_ID_UINT8CLAMPEDARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_INT16_ARRAY_UL, ECMA_BUILTIN_ID_INT16ARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_UINT16_ARRAY_UL, ECMA_BUILTIN_ID_UINT16ARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_INT32_ARRAY_UL, ECMA_BUILTIN_ID_INT32ARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_UINT32_ARRAY_UL, ECMA_BUILTIN_ID_UINT32ARRAY)
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_FLOAT32_ARRAY_UL, ECMA_BUILTIN_ID_FLOAT32ARRAY)
  #if CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64
      TYPEDARRAY_ID_CASE (LIT_MAGIC_STRING_FLOAT64_ARRAY_UL, ECMA_BUILTIN_ID_FLOAT64ARRAY)
  #endif /* CONFIG_ECMA_NUMBER_TYPE == CONFIG_ECMA_NUMBER_FLOAT64 */
      default:
      {
        JERRY_UNREACHABLE ();
      }
    }
  #undef TYPEDARRAY_ID_CASE
} /* ecma_typedarray_helper_get_builtin_id */
``` 
(Note: this function is the opposite of `ecma_typedarray_helper_get_magic_string (uint8_t builtin_id)`)

I hope it solves your problem!

Reply:
Thanks Robert, I found your comment very helpful.